### ISSUE TITLE
OpenCollada Max Plugin crashes 3DS Max 2018

### ISSUE THREAD

**Issue by ragerma** [KhronosGroup/OpenCOLLADA#540]
I tried to compile a Max plugin for 3DS Max 2018 with Visual Studio 2017.  But the plugin crashes in Options::ExportOptionsDlgProc() due to memory access violation. Does anyone have the same issue?

**Comment by RemiArnaud** [comment]
2018 is not available on our Autodesk Khronos portal. I assume it is only available in preview to a few only? or in Beta?

**Comment by ufoxp** [comment]
Educational portal has 2018 for a few months now. Same with Trial. I can confirm that even building with new SDK results in crashes. I do not see where crash comes from yet but it seems to be something else than Options::ExportOptionsDlgProc. I do not have full debug version of SDK so it will / would be hard to narrow down anyways...

**Comment by ufoxp** [comment]
On my PC removing case 1 & 2 from LibClassDesc allowwed plugin to load but then it crashes before export dialog is shown. Probably what original poster saw...

**Comment by ufoxp** [comment]
I've found one problem:
SetWindowLongPtr(hWnd, GWLP_USERDATA, (LONG)lParam);
lParam gets truncated... cast should be to LONG_PTR...

Unfortunately that is not all. COLLADAMax::getGoogleWarehouseDesc() crashes 3DSMAX on startup and something else crashes as soon as you export anything else but an empty scene.

**Comment by ufoxp** [comment]
I've made exporter work (message box based debugging for the win...) but I do not know how to change
GeometryExtra::exportParamBlock2. pblock2 is now a protected field so no way to directly access it other than casting it to some hack class which would have accessor for it. Any idea what is up with that? If I knew what to do with this I could go on to fix Google Warehouse component or submit exporter fixes without fixing warehouse.

**Comment by ufoxp** [comment]
RefTargetHandle reference = ( ( SimpleObject2 * ) mObject )->GetReference(0);
IParamBlock2 * paramBlock = ( IParamBlock2 * ) reference;
did the trick but it looks like in SDK 2018 exportParamBlock became useless. At least for spheres and cylinders now exportParamBlock2 should be used? Should I branch using macros or try to change to new way and you will see if older 3DSMAXes work with that?

**Comment by ragerma** [comment]
I used the following code, which works for older versions:
#ifdef MAX_2018_OR_NEWER
        IParamBlock2 * paramBlock = ( ( SimpleObject2 * ) mObject ) ->GetParamBlock(0);
#else
		IParamBlock2 * paramBlock = ((SimpleObject2 *)mObject)->pblock2;
#endif

**Comment by ufoxp** [comment]
Yeah, I definately could do that but I need opinion of maintainers. Also not sure when exportParamBlock stopped working? Can somebody test on 2017?

**Comment by RemiArnaud** [comment]
I prepared a CI build including Max2018. 
The build fails for 2018. Please see last message in https://github.com/KhronosGroup/OpenCOLLADA/pull/543#issuecomment-324156851

Please get a clone of that branch and propose a fix as a new PR. Jenkins will be able to build it. 

Thanks

**Comment by RemiArnaud** [comment]
I added your suggested patch in the PR branch. It's still not building release for some other reason.
best is to fix 2018 build, and then we can look into other issues separately. Building all versions sucessfully is a requirement for doing any release.

**Comment by RemiArnaud** [comment]
BTW - you are supposed to use 2015 update 3 for compiling Max 2018 plugins, not VS 2017

https://help.autodesk.com/view/3DSMAX/2018/ENU/?guid=__developer_about_the_3ds_max_sdk_sdk_requirements_html

**Comment by RemiArnaud** [comment]
@ragerma @ufoxp - the CI system has been updated, the branch / source modified to include Max 2018, Max 2018 SDK, Visual 2015 update 3, windows SDK installed. 
Also the patch you propose here was added to the code.
The artifacts are available for download from the Bot comment.

Could you please let me know if the plugin works to you satisfaction ?

**Comment by ufoxp** [comment]
I've tested from source build... As I said above this is definitely not enough to make this work. Google Warehouse and Importers are still crashing plugin startup. Also last time I tested most exportParamBlock in GeometryExtra::doExport() seemed to need changing to exportParamBlock2, otherwise extra was not exported.

BTW what is up with SDK env variable? It is named differently on my system. I've noticed that a long time ago that I always have to change those to make things build. Also post build event for plugin has path for 2017...

**Comment by ufoxp** [comment]
...and you did not fix:
SetWindowLongPtr(hWnd, GWLP_USERDATA, (LONG)lParam);

**Comment by RemiArnaud** [comment]
@ufoxp you can clone the project in your own github account. Then you can work on that branch and push changes on your own GitHub clone. Then you can open a PR here, proposing the changes from your branch to be applied to he master in Khronos/opencollada.  
From there I can invoke the jenkins build. 

Note that currently all release builds as blocked to having at least 2018 not breaking the build. So as priority it is important to merge the build changes. 

As far as the google warehouse, I don't consider this as a core feature, so I see no problem getting rid of the entire feature in order to get the import/export to work.

**Comment by RemiArnaud** [comment]
The tmp file is gone already, I can rewrite history removing the commit message, but that usually not recommended for branches that are already in a PR. You can rewrite history in your own branch. 
The project files are modified inside visual studio. Note that they have to be visual 2015 for 2018 and for the entire .sln file.  Reference to older visual studio versions are done in the sln project and need to start as is (do not auto update the solution to a different version of visual studio)

**Comment by RemiArnaud** [comment]
The SDK env is specified in this document https://github.com/KhronosGroup/OpenCOLLADA/blob/master/COLLADAMax/BUILD

With experience we learned we cannot rely on products to stay consistent, so we have our own. Read about the recent issue about visual studio 2017 env variable for instance!

**Comment by ufoxp** [comment]
I do know how to contribute in GitHub ;) Somehow I did not read "Please get a clone of that branch and propose a fix as a new PR" to the end and thought that I should push to that branch :) Heh.

This env variable thing is a little inconsistent given that you do not have own var for 3DS MAX location (where plugins land) but I guess it is of lower priority given that this mostly for debugging not CI builds...

**Comment by RemiArnaud** [comment]
I have made the two changes you called for in the branch
I will start another build (takes a while to complete as it builds all versions, all platforms, test...)

**Comment by ufoxp** [comment]
No need. I have a PR with those open already.

**Comment by ufoxp** [comment]
My changes should fix most serious crashes. At least it is functional in my use cases now without commenting out code.

**Comment by ufoxp** [comment]
I've tested 3DS MAX 2018 artifacts above and it works. I've also unblocked 3DS MAX in firewall and Google Warehouse home page URL in plugin code appears to be obsolete. Did not test further. It still crashes when closed, no surprise there...

**Comment by RemiArnaud** [comment]
closing this issue now that the PR has merged and 2018 is available https://github.com/KhronosGroup/OpenCOLLADA/releases/tag/v1.6.59

please open new issues for max plugin bugs
